Fix Windows Update CSP not deploying via GitOps

tux234 opened the pull request:
When Windows OS update settings (deadline_days and grace_period_days) are configured via GitOps YAML, the configuration is saved but the Windows Update CSP profiles are never deployed to enrolled devices.

Root cause: The GitOps code path (editTeamFromSpec) was missing the deployment trigger that exists in the UI path (ModifyTeam).

This fix adds the missing deployment logic following the same pattern used by macOS/iOS/iPadOS updates:

1. Track when Windows updates are edited (mdmWindowsUpdatesEdited flag)
2. After SaveTeam(), trigger CSP deployment via mdmWindowsEnableOSUpdates() or mdmWindowsDisableOSUpdates() based on whether deadline is set

The fix ensures GitOps and UI paths behave consistently for Windows OS update management.

Fixes: Windows Update CSP deployment via GitOps
Related: ee/server/service/teams.go lines 379-408 (UI implementation)

<!-- Add the related story/sub-task/bug number, like Resolves #123, or remove if NA -->
**Related issue:** Resolves #37137

# Checklist for submitter

If some of the following don't apply, delete the relevant line.

- [ ] Changes file added for user-visible changes in `changes/`, `orbit/changes/` or `ee/fleetd-chrome/changes`.
  See [Changes files](https://github.com/fleetdm/fleet/blob/main/docs/Contributing/guides/committing-changes.md#changes-files) for more information.

- [ ] Input data is properly validated, `SELECT *` is avoided, SQL injection is prevented (using placeholders for values in statements)
- [ ] If paths of existing endpoints are modified without backwards compatibility, checked the frontend/CLI for any necessary changes

## Testing

- [x] Added/updated automated tests
- [ ] Where appropriate, [automated tests simulate multiple hosts and test for host isolation](https://github.com/fleetdm/fleet/blob/main/docs/Contributing/reference/patterns-backend.md#unit-testing) (updates to one hosts's records do not affect another)

- [ ] QA'd all new/changed functionality manually

For unreleased bug fixes in a release candidate, one of:

- [ ] Confirmed that the fix is not expected to adversely impact load test results
- [ ] Alerted the release DRI if additional load testing is needed

## Database migrations

- [ ] Checked schema for all modified table for columns that will auto-update timestamps during migration.
- [ ] Confirmed that updating the timestamps is acceptable, and will not cause unwanted side effects.
- [ ] Ensured the correct collation is explicitly set for character columns (`COLLATE utf8mb4_unicode_ci`).

## New Fleet configuration settings

- [ ] Setting(s) is/are explicitly excluded from GitOps

If you didn't check the box above, follow this checklist for GitOps-enabled settings:

- [ ] Verified that the setting is exported via `fleetctl generate-gitops`
- [ ] Verified the setting is documented in a separate PR to [the GitOps documentation](https://github.com/fleetdm/fleet/blob/main/docs/Configuration/yaml-files.md#L485)
- [ ] Verified that the setting is cleared on the server if it is not supplied in a YAML file (or that it is documented as being optional)
- [ ] Verified that any relevant UI is disabled when GitOps mode is enabled

## fleetd/orbit/Fleet Desktop

- [ ] Verified compatibility with the latest released version of Fleet (see [Must rule](https://github.com/fleetdm/fleet/blob/main/docs/Contributing/workflows/fleetd-development-and-release-strategy.md))
- [ ] If the change applies to only one platform, confirmed that `runtime.GOOS` is used as needed to isolate changes
- [ ] Verified that fleetd runs on macOS, Linux and Windows
- [ ] Verified auto-update works from the released version of component to the new version (see [tools/tuf/test](../tools/tuf/test/README.md))

tux234 committed:
Fix Windows Update CSP not deploying via GitOps

When Windows OS update settings (deadline_days and grace_period_days)
are configured via GitOps YAML, the configuration is saved but the
Windows Update CSP profiles are never deployed to enrolled devices.

Root cause: The GitOps code path (editTeamFromSpec) was missing the
deployment trigger that exists in the UI path (ModifyTeam).

This fix adds the missing deployment logic following the same pattern
used by macOS/iOS/iPadOS updates:

1. Track when Windows updates are edited (mdmWindowsUpdatesEdited flag)
2. After SaveTeam(), trigger CSP deployment via mdmWindowsEnableOSUpdates()
   or mdmWindowsDisableOSUpdates() based on whether deadline is set

The fix ensures GitOps and UI paths behave consistently for Windows
OS update management.

Fixes: Windows Update CSP deployment via GitOps
Related: ee/server/service/teams.go lines 379-408 (UI implementation)

tux234 committed:
Add missing mock functions for Windows Update CSP deployment in tests

The TestApplyTeamSpecs and TestGetTeamsYAMLAndApply tests were failing
with nil pointer dereference when our new Windows Update deployment code
path was triggered.

The tests' mock datastore was missing the SetOrUpdateMDMWindowsConfigProfileFunc
and DeleteMDMWindowsConfigProfileByTeamAndNameFunc initializations, causing
a panic when mdmWindowsEnableOSUpdates() tried to call them.

This fix adds the missing mock function initializations to match the
pattern used in other tests (TestApplyTeamSpecsRoundtrip,
TestNoAgentOptionsCommandForPremiumUsers, etc.).

tux234 committed:
Add Windows Update CSP mock functions to all fleetctl tests

Fixed remaining test failures by adding SetOrUpdateMDMWindowsConfigProfileFunc
and DeleteMDMWindowsConfigProfileByTeamAndNameFunc mock initializations to:

- get_test.go (TestGetTeamsYAMLAndApply): 1 location
- gitops_test.go (TestGitOpsBasicTeam, TestGitOpsFullTeam, TestGitOpsBasicGlobalAndTeam, etc.): 7 locations
- apply_test.go (already fixed in previous commit): 4 locations

These mocks are required because our Windows Update CSP deployment code
in editTeamFromSpec() now calls mdmWindowsEnableOSUpdates() and
mdmWindowsDisableOSUpdates(), which need these datastore functions.

tux234 committed:
Add Windows CSP mock functions to shared test utilities

tux234 committed:
Merge branch 'main' into tux234-fix-windows-gitops-csp

tux234 committed:
Merge branch 'main' into tux234-fix-windows-gitops-csp

MagnusHJensen reviewed (commented):
@tux234 Could we add a case to maybe both gitops_test and apply_test, that covers and verifies this new behaviour is called with certain inputs, such as what happens if left empty, what if empty values "", and with values.

tux234 committed:
Add test coverage for Windows OS updates in GitOps and apply workflows

Add comprehensive test cases to verify Windows OS updates profile
management behavior in both GitOps and apply workflows.

Tests verify three scenarios:
- With values: SetOrUpdateMDMWindowsConfigProfile called with profile data
- With null values: DeleteMDMWindowsConfigProfileByTeamAndName called to disable
- Field omitted: Neither function called (no-op)

This addresses PR feedback requesting verification that the new
Windows updates logic correctly calls enable/disable functions based
on optjson.Int Set/Valid states.

tux234 committed:
Fix missing mock functions in TestGitOpsWindowsUpdates

Add required mock functions that were causing nil pointer dereference:
- GetLabelSpecsFunc
- DeleteIconsAssociatedWithTitlesWithoutInstallersFunc
- SetTeamVPPAppsFunc
- BatchInsertVPPAppsFunc
- TeamByFilenameFunc
- TeamLiteFunc
- DefaultTeamConfigFunc
- SaveDefaultTeamConfigFunc
- setupDefaultTeamConfigMocks

These mocks are required by the GitOps workflow when processing team specs.

tux234 committed:
Fix TestGitOpsWindowsUpdates YAML spec validation errors

Add required GitOps YAML fields to test specs:
- controls, queries, policies, agent_options, software
- team_settings.secrets

Add missing mock functions:
- ApplyEnrollSecretsFunc
- BatchSetSoftwareInstallersFunc
- BatchSetInHouseAppsInstallersFunc
- GetSoftwareInstallersFunc
- ListSoftwareTitlesFunc
- SetOrUpdateMDMAppleDeclarationFunc
- DeleteMDMAppleDeclarationByNameFunc
- DeleteSetupExperienceScriptFunc
- LabelIDsByNameFunc

GitOps workflow validates complete team spec structure.

tux234 committed:
Merge branch 'main' into tux234-fix-windows-gitops-csp

tux234 committed:
Deploy Windows updates CSP on team create

Trigger Windows OS updates profile creation when creating a team via GitOps, and fix gofmt in apply tests.

tux234 committed:
Fix GitOps Windows updates test YAML

Use controls.windows_updates in TestGitOpsWindowsUpdates to match GitOps schema.

tux234 committed:
Merge branch 'main' into tux234-fix-windows-gitops-csp

MagnusHJensen commented on ee/server/service/teams.go:
```diff
@@ -1589,6 +1591,17 @@ func (svc *Service) editTeamFromSpec(
 		}
 	}
+	if mdmWindowsUpdatesEdited {
+		if team.Config.MDM.WindowsUpdates.DeadlineDays.Valid {
```
Actually seems to also trigger on missing values, which is the more correct approach. I think this is okay.

MagnusHJensen reviewed (approved):
LGTM, thank you for the fix!
